go-controller/pkg/clustermanager/status_manager/zone_tracker/zone_tracker.go

```diff
@@ -0,0 +1,239 @@
+		return nil
+	}
+
+	zoneDeleted := zt.deleteNodeZoneNoLock(nodeName)
+	// notify subscriber if the zone was deleted or added (zone is added if it wasn't in the zones set, and we are adding it in 2 lines)
+	notifySubscriber := zoneDeleted || !zt.zones.Has(zoneName)
+
+	zt.nodeToZone[nodeName] = zoneName
+	zt.zones.Insert(zoneName)
+
+	if zoneName == UnknownZone {
+		// wait for unknownZoneTimeout to allow node to get a zone label.
+		// if that doesn't happen, consider a node is not a part of any zone and remove
+		go func() {
+			select {
+			case <-zt.stopChan:
+				return
+			case <-time.After(zt.unknownZoneTimeout):
+				// if node still doesn't have a zone, delete
+				zt.zonesLock.Lock()
+				defer zt.zonesLock.Unlock()
+
+				if zt.nodeToZone[nodeName] == UnknownZone {
+					klog.Warningf("Node %s has UnknownZone for more than %v", nodeName, zt.unknownZoneTimeout)
+					if zt.deleteNodeZoneNoLock(nodeName) {
+						zt.notifySubscriber()
+					}
+					zt.unknownZoneNodes.Insert(nodeName)
+				}
+			}
+		}()
+	}
+
+	if notifySubscriber {
+		zt.notifySubscriber()
+	}
+	return nil
+}
+
+func (zt *ZoneTracker) deleteNode(nodeName string) {
+	zt.zonesLock.Lock()
+	defer zt.zonesLock.Unlock()
+
+	if zt.deleteNodeZoneNoLock(nodeName) {
+		zt.notifySubscriber()
+	}
+}
+
+// deleteNodeZoneNoLock deletes a node and its zone if there are no nodes referencing it without acquiring a lock.
+// return true if a zone was deleted, must be called with zonesLock
+func (zt *ZoneTracker) deleteNodeZoneNoLock(nodeName string) bool {
+	// in case node is in the UnknownZone, delete
+	zt.unknownZoneNodes.Delete(nodeName)
+
+	oldZone, ok := zt.nodeToZone[nodeName]
+	if !ok {
+		// node doesn't exist in the map
+		return false
+	}
+	delete(zt.nodeToZone, nodeName)
+	deleteZone := true
+	for _, existingZone := range zt.nodeToZone {
+		if existingZone == oldZone {
+			// at least one node in the same zone exists, don't delete
+			deleteZone = false
+			break
+		}
+	}
+
+	if !deleteZone {
+		return false
+	}
+
+	zt.zones.Delete(oldZone)
+	return true
+}
+
+// Subscribe allows to get notified about zones changes (not nodes changes,
+// e.g. if a new node for existing zone is added, callbacks won't be executed).
+// First callback will be executed when zoneTracker finishes its initial sync, that may happen immediately.
+// Only 1 subscriber is allowed for now, therefore an error will be returned on the second call.
+func (zt *ZoneTracker) Subscribe(onZonesUpdate func(newZones sets.Set[string])) error {
+	zt.zonesLock.Lock()
+	defer zt.zonesLock.Unlock()
+	if zt.onZonesUpdate != nil {
+		return fmt.Errorf("ZoneTracker allows only 1 subscriber")
+	}
+	zt.onZonesUpdate = onZonesUpdate
+	zt.notifySubscriber()
+	return nil
+}
+
+// notifySubscriber must be called with zonesLock
+func (zt *ZoneTracker) notifySubscriber() {
+	if zt.inited && zt.onZonesUpdate != nil {
+		zt.onZonesUpdate(sets.New[string](zt.zones.UnsortedList()...))
+	}
+}
```

Comment:
This is just a nit, but if there can ever be just one subscriber, why just not pass the user callback in the zone tracker constructor. Otherwise the double indirection is just confusing and seems more complicated that needs be